### PR TITLE
Bag of words

### PR DESCRIPTION
**Pull request by JinfengXiao**

### CODE REVIEW

**Review by mattvitae — CHANGES_REQUESTED** [review]
Thank you @JinfengXiao for this contribution!

I have requested some changes to make this code more generalized and more maintainable in the future, so that we can easily contribute to the code and work with you more easily.

The details are in the comments within this review.

File: README.md [hunk 1]
```diff
@@ -19,3 +19,13 @@ To load test documents into the index, start up the cluster and run:
 $ test_index.sh
 ```
 
+## Bag of Words (BoW)
+
+Note: Only codes are included. Data files are not.
```
Comment (mattvitae) [inline README.md]: Can you please include a small sample of the data files for testing (I would like help with code modifications, but I don't want to break what you've done!)

File: bag_of_words/code/bag_of_words.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# Read papers and convert them to bag-of-words representations.
```
Comment (mattvitae) [inline bag_of_words/code/bag_of_words.py]: I recommend that we try to generalize this code so that we can easily choose a different vectorized (such as TF-IDF) as an argument to the code, or perhaps a different function called for a different vectorizer. I will try to leave some example snippets below.

File: bag_of_words/code/bag_of_words.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# Read papers and convert them to bag-of-words representations.
+
+indir = "../data/raw/pubmed-txts-mf/"
```
Comment (mattvitae) [inline bag_of_words/code/bag_of_words.py]: I recommend following [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines for Python code style. 

Here are some examples:
- please use CAPITAL_LETTERS for constant var names.
- please put `import` statements first in the code
- please do not initialize objects outside of a function (or method, if you use classes)

For code that can be called as a module or stand-alone, please use the `if __name__ == "__main__"` convention.

File: bag_of_words/code/bag_of_words.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+indir = "../data/raw/pubmed-txts-mf/"
+outdir = "../data/processed/"
+from sklearn.feature_extraction.text import CountVectorizer
+vectorizer = CountVectorizer()
+import os
+import numpy as np
+import pickle
+
+# Read papers into a list
```
Comment (mattvitae) [inline bag_of_words/code/bag_of_words.py]: I recommend writing this section as two functions: `read_corpus` and `tokenize_corpus`,  and then using an option arg to `tokenize_corpus` to write the corpus to either `csv` or `pkl` formats (or none at all).

File: bag_of_words/code/bag_of_words.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+pmids = os.listdir(indir)
+pmids.sort()
+# pmids = pmids[:5]  # For debugging
+corpus = [''] * len(pmids)
```
Comment (mattvitae) [inline bag_of_words/code/bag_of_words.py]: Since we end up iterating over the PMID to build the corpus, and indexing by the PMID, I recommend using a dict of strings here instead of a list. That way, you can refer to the PMID as the dict key, and you don't need to keep a separate list of PMID.

File: bag_of_words/code/bag_of_words.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+print("Reading corpus ...")
+for i in range(len(pmids)):
```
Comment (mattvitae) [inline bag_of_words/code/bag_of_words.py]: instead, you could have `for pmid in pmids: ... `

File: bag_of_words/code/bag_of_words.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    if i % 10000 == 0:
+        print(str(i) + " docs processed ...")
+    for field in os.listdir(indir + pmids[i] + '/'):
+        with open(indir + pmids[i] + '/' + field, 'r') as fin:
+            text = fin.read().replace('\n', ' ')
+            corpus[i] += text
+print('')
```
Comment (mattvitae) [inline bag_of_words/code/bag_of_words.py]: I recommend using python's `logger` module, which you can simply use in files like this by using `logging.getLogger()`

File: bag_of_words/code/bag_of_words.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+# Read papers into a list
+pmids = os.listdir(indir)
+pmids.sort()
+# pmids = pmids[:5]  # For debugging
+corpus = [''] * len(pmids)
+print("Reading corpus ...")
+for i in range(len(pmids)):
+    if i % 10000 == 0:
+        print(str(i) + " docs processed ...")
```
Comment (mattvitae) [inline bag_of_words/code/bag_of_words.py]: I also recommend using newer string formatting methods for Python, e.g. `"%s docs processed".format(i)`

File: bag_of_words/code/bag_of_words.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    for field in os.listdir(indir + pmids[i] + '/'):
+        with open(indir + pmids[i] + '/' + field, 'r') as fin:
+            text = fin.read().replace('\n', ' ')
+            corpus[i] += text
+print('')
+
+# Convert papers to bags of words, and save objects to file
+print("Building bag-of-words representation ...")
+bow = vectorizer.fit_transform(corpus)
```
Comment (mattvitae) [inline bag_of_words/code/bag_of_words.py]: If this were a `def vectorize_corpus` function, it could default to `CountVectorizer` to make the `bow`, but it could also accept a vectorizer object as an argument, in which case, you could simply call the `bow` variable something like `vectors`, which would generalize to TF-IDF or other tokenizers and vectorizers.

File: bag_of_words/code/bag_of_words.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+with open(outdir + "bag_of_words.pkl", 'wb') as fout:
+    pickle.dump(bow, fout)
+    bow_rownames = pmids
+    pickle.dump(bow_rownames, fout)
+    bow_colnames = vectorizer.get_feature_names()
+    pickle.dump(bow_colnames, fout)
```
Comment (mattvitae) [inline bag_of_words/code/bag_of_words.py]: I recommend logging when you save the file. And if you have a function for this section, it can accept an arg for the `pkl` filename, or even to save in different formats (you don't have to implement them all now, but allowing for the future extension is very good practice).

File: bag_of_words/code/classifiers.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+from sklearn.model_selection import cross_validate
+from sklearn.svm import LinearSVC
+from sklearn.tree import DecisionTreeClassifier
+from sklearn.ensemble import RandomForestClassifier
+
+indir = "../data/processed/"
+
+with open(indir + "bag_of_words.pkl", 'rb') as fin:
+    bow = pickle.load(fin)  # doc-by-word count sparse matrix
+    bow_rownames = pickle.load(fin)  # list of PMIDs
```
Comment (mattvitae) [inline bag_of_words/code/classifiers.py]: It's not clear to me how you're getting rownames and colnames separately from this `pickle.load`

File: bag_of_words/code/classifiers.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    bow_colnames = pickle.load(fin)  # list of distinct words
+
+with open(indir + "doc_labels.pkl", 'rb') as fin:
+    labels = pickle.load(fin)  # doc labels
+
+# Remove docs that have no labels
```
Comment (mattvitae) [inline bag_of_words/code/classifiers.py]: This would be good to have as a separate function.

File: bag_of_words/code/classifiers.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+labels_pos_ind = (labels.sum(axis=1) > 0)
+labels_pos_ind.sum() * 1.0 / len(labels_pos_ind)
+bow = bow[labels_pos_ind.values,:]
+bow_rownames = list(compress(bow_rownames, labels_pos_ind.values))
+labels = labels[labels_pos_ind.values]
+
+# Take the topk tokens
```
Comment (mattvitae) [inline bag_of_words/code/classifiers.py]: This would be good as a function, with arguments for the topk arg, for example.

File: bag_of_words/code/classifiers.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+topk = 1000
+df = (bow > 0).sum(axis=0)
+bow = bow[:,(df > 0).A1]  # Remove words with zero occurance in the sub-corpus
+df = (bow > 0).sum(axis=0)
+tf = bow.sum(axis=0)
+idf = np.log( (len(bow_rownames) + 1) / df )
+tfidf = np.multiply(tf, idf)
```
Comment (mattvitae) [inline bag_of_words/code/classifiers.py]: I recommend that we use the TF-IDF vectorizers, which are optimized, and also will ensure we don't make any simple mistakes with these calculations.

Comment (mattvitae) [reply bag_of_words/code/classifiers.py]: Those vectorizers also come with built-in methods for selecting cutoffs.

File: bag_of_words/code/classifiers.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+# Select tokens based on TF-IDF
+cutoff = heapq.nlargest(topk, tfidf.A1.tolist())[topk-1]
+bow_freq = bow[:,(tfidf >= cutoff).A1]
+# Select tokens based on DF
+#cutoff = heapq.nlargest(topk, df.A1.tolist())[topk-1]
+#bow_freq = bow[:,(df >= cutoff).A1]
+
+# Proportion of False labels (i.e. accuracy of an all-False classifier)
```
Comment (mattvitae) [inline bag_of_words/code/classifiers.py]: I would prefer a `majority class` classifier as our weak baseline.

File: bag_of_words/code/classifiers.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+#false_prop.sort_values()
+
+
+# Train and test classifiers for a specific label
+cv_fold = 3
+for label in ["individual_observations", "functional_experiments", "family_studies", "sequence_observations"]:
+    print("Class: " + label)
+    scoring = ["precision", "recall", "f1", "accuracy"]
+    
+    print("LinearSVC with L2:")
```
Comment (mattvitae) [inline bag_of_words/code/classifiers.py]: By theory, we should prefer L1 for discrete spaces like Count, and L2 for continuous spaces like TF-IDF

File: bag_of_words/code/classifiers.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+#cutoff = heapq.nlargest(topk, df.A1.tolist())[topk-1]
+#bow_freq = bow[:,(df >= cutoff).A1]
+
+# Proportion of False labels (i.e. accuracy of an all-False classifier)
+false_prop = (labels == False).sum(axis=0) / labels.shape[0]
+#false_prop.sort_values()
+
+
+# Train and test classifiers for a specific label
+cv_fold = 3
```
Comment (mattvitae) [inline bag_of_words/code/classifiers.py]: I recommend making this a function that takes args for `cv_fold` as well as the `label` we want to classify, the scoring metrics, and the models we're interested in running.

File: bag_of_words/code/doc_labels.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
+# Extract the labels of documents, and store them into an array.
```
Comment (mattvitae) [inline bag_of_words/code/doc_labels.py]: Similar to the other files, I recommend writing functions and following PEP8 for this code.

**Review by mattvitae — CHANGES_REQUESTED** [review]
@JinfengXiao Thank you again for this contribution! We are on our way :-)

I think we should begin storing the documents in the Elastic Search index with a document schema, and then you will not have to read in the raw text files. Instead, we can retrieve them from the index, and modify their classifications with attributes in the index. Of course, I know that we're very early on and still experimenting, I only want to suggest that we should be thinking about how this code will run in a production system before we get too far with scripts like these. We can help you with this reorganization.

File: README.md (hunk 1 above)
Comment (JinfengXiao) [reply README.md]: Yes! I added a sample dataset to the commit on 10/11/2018.

File: bag_of_words/code/bag_of_words.py (hunk 2 above)
Comment (JinfengXiao) [reply bag_of_words/code/bag_of_words.py]: I tried to rewrite the code into functions. Could you have a look?

File: bag_of_words/code/bag_of_words.py (hunk 3 above)
Comment (JinfengXiao) [reply bag_of_words/code/bag_of_words.py]: I have enhanced the coding style of all scripts except classifiers.py which is under active experimentation.

File: bag_of_words/code/bag_of_words.py (hunk 4 above)
Comment (JinfengXiao) [reply bag_of_words/code/bag_of_words.py]: solved

File: bag_of_words/code/classifiers.py (hunk 11 above)
Comment (JinfengXiao) [reply bag_of_words/code/classifiers.py]: I think calling pickle.load(fin) for 3 times will read in 3 objects in the order that they were written.

**Review by mattvitae — APPROVED** [review]
